### DISCUSSION TITLE
Why the word "[Theory]" as opposed to something like "[MultiFact]"?

### DISCUSSION THREAD

**Discussion by abergquist** [xunit/xunit#2822] [Question] [answered]
We've been using **xUnit** for testing our **.NET MAUI** code. It's been great!

I have to ask, though; why the attribute "[Theory]"? It seems to me that something like "[MultiFact]" or "[TestCases]" is more mnemonic.

**Comment by bartelink** [comment]
It's just naming; The names are accentuating the role it plays vs a lower level description of what it technically does/is. Whether that's a win is subjective; you can argue it puts authors in the right frame of mind. Or you can argue that it's cryptic. But xUnit is definitely designed with lots of though put into nuanced design aspects with a view to nudging one towards writing good tests, and shaping good systems.

In NUnit
We are testing so everything is a Test
And a Test that runs many times is a MultiTest
And the data comes from a TestCaseData or a TestCaseSource

In xUnit
A Fact is an individual relevant assertion. If you have 3 of them, they may or may not mean anything in aggregate.
A Theory is a set of facts that together prove an overall concept
XData can be an individual case, or an aggregate source (coz Data can be singular or plural)

In Property Based Testing, a Property is called that because that's its role - It's not a TestWithMultipleInvocationsBasedOnArbitraryData. We are specifting a property of the system under test

**Comment by bradwilson** [comment, marked as answer]
From https://xunit.net/docs/why-did-we-build-xunit-1.0:

> The definition of how to run a test method can be extended. There are two example of this: the first, in xunitext.dll, is the [Theory] attribute which allows data-driven tests; the second, in the samples, is the [RepeatTest] attribute which runs a test method multiple times in a row. For more information on data theories, see https://homes.cs.washington.edu/~mernst/pubs/testing-theories-tr002-abstract.html.

**Reply by bradwilson** [reply]
We had always planned to have theories (they're now an integral part of the framework and not just in the now-defunct extension system). That's why we used `[Fact]` instead of `[Test]` as well, because of the language symmetry.

When asked for the elevator pitch on the difference, we would say "facts are tests whose code is always true; theories are tests whose code is only conditionally true, depending on the input data" with the canonical example of a theory here being something that checks a function for determining whether a integral value is odd or even. You write the "odd" test using only odd values, and know that the test would (incorrectly) fail if you passed it an even value.

A shorter version might be "facts test invariants; theories test conditions".

**Reply by abergquist** [reply]
@bartelink and @bradwilson I really appreciate all of your input, insights and feedback regarding this question.

I'm always fine with learning new nomenclature and concepts but for better grokking by the vast majority of developers, I'll think of them as **[UnitTest]** and **[DataDrivenTests]** (or **[Fact]** and **[FactSet]**), my rationale being the principle of naming concepts with the clearest (read: least esoteric) name as possible.

**Reply by bradwilson** [reply]
You're welcome to write your own attributes that derive from FactAttribute and TheoryAttribute, if that makes things easier. 😄

**Reply by bartelink** [reply]
I can well appreciate that the naming approach doesn't appeal to all; it's the ultimate subjective thing.

The name Theory makes complete sense to me (but I also appreciate the wonderful https://en.wikipedia.org/wiki/Thing_Explainer)

**Reply by abergquist** [reply]
Thanks @bartelink.

I believe the book mentioned within that link is now going to become a Christmas present! 

Thanks again!

**Reply by bartelink** [reply]
It's fantastic (as are his 'what if' books - they're fantastic for adults too, but I also have one kid that adores them; there's a website that's representative of the book content if you want a preview...)

**Reply by abergquist** [reply]
You bet I want a preview; thanks!

**Reply by bartelink** [reply]
https://what-if.xkcd.com/

**Reply by abergquist** [reply]
Thanks, @bartelink!
